agent/discover/discover.go

```diff
@@ -0,0 +1,52 @@
+package discover
+
+import (
+	"fmt"
+	"os"
+
+	consul "github.com/hashicorp/consul/api"
+	consul_internal "github.com/trento-project/trento/internal/consul"
+)
+
+type Discovery func() (Discoverer, error)
+
+type Discoverer interface {
+	// this function checks whether this particular implementation of Discover
+	// is relevant for this node. It is used as a gating condition for other
+	// functionality in this implementation
+	ShouldDiscover(client consul.Client) bool
+	// Execute one iteration of a discovery and store the result in the Consul
+	// KVStore.
+	Discover() error
+
+	// Create or Updating the given Consul Key-Value Path Store with a new value from the Agent
+	storeDiscovery(cStorePath, cStoreValue string) error
+}
+
+type Discover struct {
+	client consul.Client
+	host   string
+}
+
+// Create or Updating the given Consul Key-Value Path Store with a new value from the Agent
+func (discover Discover) storeDiscovery(cStorePath, cStoreValue string) error {
+	kvPath := fmt.Sprintf("%s/%s/%s", consul_internal.KvHostsPath, discover.host, cStorePath)
+
+	_, err := discover.client.KV().Put(&consul.KVPair{
+		Key:   kvPath,
+		Value: []byte(cStoreValue)}, nil)
+	return err
+}
+
+// Execute one iteration of a discovery and store the result in the Consul KVStore.
+func (discover Discover) Discover() error {
+	return nil
+}
+
+// Return a Host Discover instance
+func NewDiscover(client consul.Client) Discover {
+	r := Discover{}
+	r.client = client
+	r.host, _ = os.Hostname()
```

Comment:
hmm, shouldn't this be done in the `Discover` function? Not that it's likely for the hostname to change, but still, it could happen without a reboot, right?

Reply:
I duplicated it into the Discover() function as well, as it was empty so far. I need it set on construction just to make sure that storeDiscovery() calls are ending up in the right spot.